translated/tech/201711 Golang Pros and Cons for DevOps/Goroutines, Panics, and Errors.md

```diff
@@ -0,0 +1,470 @@
+func monitorNetwork() { … }
+func monitorProcesses() { … }
+func monitorIdentity() { … }
+
+func main() {
+	for !shutdown {
+		monitorCpu()
+		monitorDisk()
+		monitorNetwork()
+		monitorProcesses()
+		monitorIdentity()
+		time.Sleep(5*time.Second)
+	}
+}   
+```     
+在上面的代码中，CPU，硬盘，网络，进程和身份每 5 秒检查一次。如果上述其中之一挂掉的话，所有的监控就会终止。每一个监控用时越长，检查变得越不频繁，因为我们在计算完成之后需要睡眠 5 秒。(ps.最后一句没看懂)
+为了解决这些问题，一种解决方案(一种不完整的方案，但是完美展示 goroutine 的价值所在)是使用 goroutine 来调用每一个监控函数。仅仅在你想要以线程方式产生的函数调用前加上 `go`关键字。     
+```
+func main() {
+	for !shutdown {
+		go monitorCpu()
+		go monitorDisk()
+		go monitorNetwork()
+		go monitorProcesses()
+		go monitorIdentity()
+		time.Sleep(5*time.Second)
+	}
+}   
+```
+现在如果它们之中任何一个挂掉的话，仅仅被阻塞的调用会被终止，不会阻塞其它的监控调用函数。并且因为产生一个线程很容易也很快，现在我们事实上更靠近了每隔 5 秒钟检查这些系统。     
+当然，上述这个解决方案也有其他问题，比如在一个 goroutine 的 panic 可能会破坏其他 goroutine，睡眠时间有少量的偏差，代码并不像看到的那样模块化等等。但是生成内核级线程不是很容易吗？    
+正如你所看到的，Java 程序需要 12 行代码，而 Go 却只需要 2 个单词。我们想说这意味着你的 Go 源代码将变得简洁和紧凑，但是当我们在这篇文章之后介绍的 panics 和 errors 时，你将会看到不幸的是并非如此(事实上 Go 语言是一种比较臃肿的语言，现在信不信由你)。    
+### 同步包(和 Channels)加上编配    
+我们使用 Go 的同步包和 channels 是为了 goroutine 的编配，发送信号和关机。你将会发现对 sync.Mutex 和 sync.WaitGroup 的引用以及一种称为 shutdownChannel 的重复的结构体变量充斥在我们的代码中。    
+在 [Go manual](https://golang.org/pkg/sync/) 中关于 sync.Mutex 和 sync.WaitGroup 有一个重要的提示:    
+>> 不应拷贝包含此包中定义的类型的值。      
+
+解释给那些整天不使用 Go，C 或者 C++ 的人(ps.拿不准):结构体是值传递的(ps.感觉应该是引用传递)。任何时间你创建一个 Mutex 或者 WaitGroup，使用一个指针，不要使用直接的值。这并不是普遍必需的，但是如果你不知道何时是好是坏，请一直使用指针。这里有一个很好的，简单的例子:    
+```
+type Example struct {
+	wg *sync.WaitGroup
+	m *sync.Mutex
+}
+
+func main() {
+	wg := &sync.WaitGroup{}
+	m := &sync.Mutex{}
+}   
+```   
+然而对于这些结构体的警告恰恰在页面的顶部，很容易掩饰过去，将会导致你将要构建的应用出现奇怪的副作用。   
+从前一节关于监控的那个例子来看，这个例子是我们如何使用一个 WaitGroup 来确保在任何时候每个系统中不会有多于一个的监控:   
+```
+import "sync"
+…
+func main() {
+	for !shutdown {
+		wg := &sync.WaitGroup{}
+
+		doCall := func(fn func()) {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				fn()
+			}
+		}
+
+		doCall(monitorCpu)
+		doCall(monitorDisk)
+		doCall(monitorNetwork)
+		doCall(monitorProcesses)
+		doCall(monitorIdentity)
+
+		wg.Wait()
+	}
+}   
+```    
+一个 mutex 是保护共享资源的好方法，比如服务器上的CPU度量历史，正在监视日志文件的水印，或对更新事件感兴趣的监听器列表。这里没有惊奇的地方，除了一个相当令人可怕的 `defer` 关键字外，但是它超出这篇文章的范畴了。   
+``` 
+package main
+
+import (
+	"fmt"
+	"sync"
+	"time"
+)
+
+func printAndSleep(m *sync.Mutex, x int) {
+	m.Lock()
+	defer m.Unlock()
+	fmt.Println(x)
+	time.Sleep(time.Second)
+}
+
+func main() {
+	m := &sync.Mutex{}
+	for i := 0; i < 10; i++ {
+		printAndSleep(m, i)
+	}
+}   
+```    
+goroutine(s) 不仅容易启动，而且总的来说它们也容易协调，关闭和等待。不过有几个复杂的事情要处理，它们可能会更加深入一些:对多线程的事件广播，工作池，还有分布式处理。   
+### Goroutines 的自动清理      
+goroutine(s) 持有堆变量和栈变量的引用(避免垃圾收集)，但是并不需要一直持有这些引用。它们(goroutines)一直运行直到函数完成，然后关闭并且自动释放所有资源。在这个过程中有一个需要注意的部分是如果主线程已经退出，启动一个 goroutine 会被忽略(或者跳过)。 
+首先，这里举一个有关启动 goroutine 并被忽略的真实例子。在我们的应用中，我们在子进程中启动一个模块(函数)，并且使用 IPC 来配置更新，设置更新和心跳。父进程和每个模块进程(子进程)必须不断地从 IPC channel 中读取数据，并且彼此向对方发送数据。这是一种我们启动并且忽略的线程，因为我们不关心如果(对方已经)关闭，我们不能读取完整的流。请持保留态度的看这段代码，虽然它来自我们的代码，但是为了简单起见，移除了一些重量级的代码:   
```

Comment:
"send that information elsewhere" 我也没有不确定这里的 elsewhere 确切指哪里，所以我建议直接就是 再向别的地方发送数据  而不是主观地翻译为：向对方发送数据 感觉不准确

Reply:
“This is the kind of thread that we launch and forget about, because we don’t care if we didn’t read the whole stream on shutdown.” 后半句翻译不准确，建议：
这是一种我们启动并且忽略的线程，因为在关机的时候，我们不关心我们是否已经接收了全部的信息流。